### PR TITLE
Updated BigQuery source in order to make code reusable for BQ Pushdown.

### PR DESCRIPTION
**Pull request by fernst**

### CODE REVIEW

File: src/main/java/io/cdap/plugin/gcp/bigquery/source/BigQueryInputFormatProvider.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  }
+
+  @Override
+  public String getInputFormatClassName() {
+    return PartitionedBigQueryInputFormat.class.getName();
+  }
+
+  @Override
+  public Map<String, String> getInputFormatConfiguration() {
+    Map<String, String> config = new HashMap<>();
```
Comment (tivv) [inline src/main/java/io/cdap/plugin/gcp/bigquery/source/BigQueryInputFormatProvider.java]: Can be done as
`StreamSupport.stream(configuration.spliterator(), false).collect(Collectors.toMap(Entry::getKey, Entry::getValue))`
Wondering why do we need to copy the whole configuration. In any case, it would be great to move to constructor.

File: src/main/java/io/cdap/plugin/gcp/bigquery/source/BigQuerySourceUtils.java [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+                            cmekKey);
+    }
+
+    return bucket;
+  }
+
+  /**
+   * Updates {@link Configuration} with bucket details.
+   */
+  public static void configureBucket(Configuration configuration, String bucket, String runId) {
```
Comment (tivv) [inline src/main/java/io/cdap/plugin/gcp/bigquery/source/BigQuerySourceUtils.java]: It looks like it's not only bucket, but a full path inside the bucket.

File: src/main/java/io/cdap/plugin/gcp/bigquery/source/BigQuerySourceUtils.java [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+    configuration.set("fs.default.name", String.format("gs://%s/%s/", bucket, runId));
```
Comment (tivv) [inline src/main/java/io/cdap/plugin/gcp/bigquery/source/BigQuerySourceUtils.java]: GCS_PATH_FORMAT?

File: src/main/java/io/cdap/plugin/gcp/bigquery/source/BigQuerySource.java [hunk 4]
```diff
@@ -82,7 +71,7 @@
   private Configuration configuration;
   private final BigQueryAvroToStructuredTransformer transformer = new BigQueryAvroToStructuredTransformer();
   // UUID for the run. Will be used as bucket name if bucket is not provided.
-  private UUID uuid;
+  private String runId;
```
Comment (tivv) [inline src/main/java/io/cdap/plugin/gcp/bigquery/source/BigQuerySource.java]: We are already using a run id term (ProgramRunId) and it looks like this guy is something else. Please use a different name

File: src/main/java/io/cdap/plugin/gcp/bigquery/source/BigQuerySourceUtils.java (hunk 3 above)
Comment (tivv) [reply src/main/java/io/cdap/plugin/gcp/bigquery/source/BigQuerySourceUtils.java]: Also, do we log this value anywhere? It makes sense to log it early to ensure we have it in case of any issues (e.g. permissions)

File: src/main/java/io/cdap/plugin/gcp/bigquery/source/BigQueryInputFormatProvider.java (hunk 1 above)
Comment (tivv) [reply src/main/java/io/cdap/plugin/gcp/bigquery/source/BigQueryInputFormatProvider.java]: I am wondering if we should switch from making Configuration that would have all the defaults, then adding few items and changing it to a map that potentially clashes with the different defaults on the other side.

Instead we should make a map and then create a Configuration only when needed (when calling org.apache.hadoop.fs.Path#getFileSystem)

File: src/main/java/io/cdap/plugin/gcp/bigquery/source/BigQuerySourceUtils.java [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+    configuration.setBoolean("fs.gs.metadata.cache.enable", false);
+  }
+
+  /**
+   * Sets up service account credentials into supplied Hadoop configuration.
+   *
+   * @param configuration Hadoop Configuration instance.
+   * @param config Big Query configuration.
+   */
+  public static void configureServiceAccount(Configuration configuration, BigQueryBaseConfig config) {
```
Comment (tivv) [inline src/main/java/io/cdap/plugin/gcp/bigquery/source/BigQuerySourceUtils.java:98]: Where is this one used?

**Review by tivv — COMMENTED** [review]
Generally I don't feel good about having this number of small methods. When using it's very easy to screw up. Also it's hard to follow which methods are called and what is set. Can we have a smaller number of bigger methods?

File: src/main/java/io/cdap/plugin/gcp/bigquery/source/BigQuerySource.java (hunk 4 above)
Comment (fernst) [reply src/main/java/io/cdap/plugin/gcp/bigquery/source/BigQuerySource.java]: This is an identifier for a given execution. Each source has it's own "runId" that is used to create a new path to store objects related to this source.

File: src/main/java/io/cdap/plugin/gcp/bigquery/source/BigQuerySourceUtils.java (hunk 5 above)
Comment (fernst) [reply src/main/java/io/cdap/plugin/gcp/bigquery/source/BigQuerySourceUtils.java:98]: Oops, will add it back.

File: src/main/java/io/cdap/plugin/gcp/bigquery/source/BigQuerySourceUtils.java (hunk 3 above)
Comment (fernst) [reply src/main/java/io/cdap/plugin/gcp/bigquery/source/BigQuerySourceUtils.java]: You're right, this should be the static variable.

File: src/main/java/io/cdap/plugin/gcp/bigquery/source/BigQuerySource.java (hunk 4 above)
Comment (tivv) [reply src/main/java/io/cdap/plugin/gcp/bigquery/source/BigQuerySource.java]: Well, it's not a key for execution, you can't retrieve an execution using it. It would be no problems if we did not have already io.cdap.cdap.common.app.RunIds widely used, but now those two will be confused and if someone try to retrieve a run using this id he/she would fail.

Comment (fernst) [reply src/main/java/io/cdap/plugin/gcp/bigquery/source/BigQuerySource.java]: I see your point now.

Comment (fernst) [reply src/main/java/io/cdap/plugin/gcp/bigquery/source/BigQuerySource.java]: I have renamed this variable to something that is less confusing.

File: src/main/java/io/cdap/plugin/gcp/bigquery/source/BigQueryInputFormatProvider.java (hunk 1 above)
Comment (yaojiefeng) [reply src/main/java/io/cdap/plugin/gcp/bigquery/source/BigQueryInputFormatProvider.java]: You can use `SourceInputFormatProvider` to just return the non-default configs, there is logic to compute the non-default here - https://github.com/cdapio/hydrator-plugins/blob/develop/hydrator-common/src/main/java/io/cdap/plugin/common/SourceInputFormatProvider.java#L40. It is in hydrator-common, so you should be able to directly use it here.

**Comment by tivv** [comment]
Generally I don't feel good about having this number of small methods. When using it's very easy to screw up. Also it's hard to follow which methods are called and what is set. Can we have a smaller number of bigger methods?

**Comment by fernst** [comment]
@tivv from the perspective of the SQL engine, I only need these 3 methods:

```
    // Configure GCS Bucket to use
    String gcsPath = BigQuerySQLEngineUtils.getGCSPath(bucket, runId, table);
    BigQuerySourceUtils.configureBucket(configuration, gcsPath);

    // Set up temporary table name. This will be used if the source table is a view
    BigQuerySourceUtils.configureTemporaryTableName(configuration, table);

    // Configure BigQuery input format.
    BigQuerySourceUtils.configureBigQueryInput(configuration, project, bqDataset, table, gcsPath);
```

That's the reason I build all of these small methods in the utility class, it allows me to bring only the small bits of functionality I need to make the SQL engine implementation work.

**Comment by fernst** [comment]
@tivv condensed some methods into one.

File: src/main/java/io/cdap/plugin/gcp/bigquery/source/BigQuerySource.java [hunk 6]
```diff
@@ -82,7 +71,7 @@
   private Configuration configuration;
   private final BigQueryAvroToStructuredTransformer transformer = new BigQueryAvroToStructuredTransformer();
   // UUID for the run. Will be used as bucket name if bucket is not provided.
-  private UUID uuid;
+  private String bucketPath;
```
Comment (tivv) [inline src/main/java/io/cdap/plugin/gcp/bigquery/source/BigQuerySource.java:74]: nit: bucketName, it's not a full path

**Review by tivv — APPROVED** [review]
